File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.java
import java.util.*;

class Solution {
    public String solution(String m, String[] musicinfos) {
        Map<String, Integer> time = new HashMap<>();
        Map<String, String> music = new HashMap<>();
        List<String> name = new ArrayList<>();
        
        for (int i = 0; i < musicinfos.length; i++) {
            String[] split = musicinfos[i].split(",");
            name.add(split[2]);
            int span = time(split[0], split[1]);
            time.put(split[2], span);
            
            String melody = music(split[3], span);
            music.put(split[2], melody);
        }

        Collections.sort(name, (a, b) -> time.get(b).compareTo(time.get(a)));
        for (int i = 0; i < name.size(); i++) {
            String melody = music.get(name.get(i));

            if (compare(m, melody)) {
                return name.get(i);
            }
        }

        return "(None)";
    }

    public int time(String start, String end) {
        int startTime = Integer.parseInt(start.split(":")[0]) * 60 + Integer.parseInt(start.split(":")[1]);
        int endTime = Integer.parseInt(end.split(":")[0]) * 60 + Integer.parseInt(end.split(":")[1]);

        return endTime - startTime;
    }

    public String music(String music, int time) {
        int idx = 0, loop = 0;
        StringBuilder sb = new StringBuilder();

        while (loop < time) {
            sb.append(music.charAt(idx));
            idx = (idx + 1) % music.length();

            if (music.charAt(idx) == '#') {
                sb.append(music.charAt(idx));
                idx = (idx + 1) % music.length();
            }

            loop++;
        }

        return sb.toString();
    }

    public boolean compare(String m, String melody) {
        int idx = 0, cnt = 0;
        while (idx < m.length()) {
            if (idx + 1 < m.length() && m.charAt(idx + 1) == '#') idx++;

            idx++;
            cnt++;
        }

        idx = 0;
        while (idx < melody.length()) {
            StringBuilder sb = new StringBuilder();

            int p = idx;
            int loop = 0;
            while (p < melody.length() && loop < cnt) {
                sb.append(melody.charAt(p++));
                if (p < melody.length() && melody.charAt(p) == '#') sb.append(melody.charAt(p++));

                loop++;
            }

            if (sb.toString().equals(m)) return true;

            idx++;
        }


        return false;
    }
}

File: 프로그래머스/2/17683. ［3차］ 방금그곡/［3차］ 방금그곡.cpp
#include <bits/stdc++.h>

using namespace std;

struct music{
    int odr;
    int st;
    int et;
    string name;
    string rhyme;
};

string solution(string m, vector<string> musicinfos) {
    string answer = "";
    string temp="";
    string time="";
    istringstream ms;
    istringstream ts;
    int rnd=0;
    int idx=0;
    int seq=1;
    int cnt=0;
    int h=0;
    int minute=0;
    int len=0;
    int sharf=0;
    bool in=0;
    vector<music> store;
    music compare;

    for(string song:musicinfos) {
        idx=0;
        compare.odr=rnd;
        ms.str(song);
        while(getline(ms,temp,',')) {
            if(idx==0) {
                seq=0;
                ts.str(temp);
                while(getline(ts,time,':')) {
                    if(seq==0) {
                        h=stoi(time);
                    }
                    else {
                        minute=stoi(time);
                    }
                    ++seq;
                }
                compare.st=h*60+minute;
                ts.clear();
            }
            else if(idx==1) {
                seq=0;
                ts.str(temp);
                while(getline(ts,time,':')) {
                    if(seq==0) {
                        h=stoi(time);
                    }
                    else {
                        minute=stoi(time);
                    }
                    ++seq;
                }
                compare.et=h*60+minute;
                ts.clear();
            }
            else if(idx==2) {
                compare.name=temp;
            }
            else {
                compare.rhyme=temp;
            }
            ++idx;
        }

        len=compare.rhyme.size();
        cnt=0;
        seq=0;
        idx=0;
        in=0;
        while(seq<=compare.et-compare.st&&cnt<m.size()) {
            
           // cout<<compare.rhyme<<", cnt:"<<cnt<<", in:"<<in<<", seq:"<<seq<<", idx:"<<idx<<", compare:"<<compare.rhyme[idx%len]<<", m:"<<m[cnt]<<endl;
            if(!in) {
                if(compare.rhyme[idx%len]==m[cnt]) {
                    in=1;
                    ++sharf;
                    ++cnt;
                }
            }
            else {
                if(compare.rhyme[idx%len]!=m[cnt]) {
                    in=0;
                    idx-=cnt;
                    seq-=(cnt-sharf+1);
                    cnt=0;
                    sharf=0;
                }
                else {
                    if(compare.rhyme[idx%len]=='#') {
                        ++sharf;
                    }
                    ++cnt;
                }
            }

            if(cnt>=m.size()&&compare.rhyme[(idx+1)%len]=='#') {
                in=0;
                idx-=(cnt-1);
                seq-=((cnt-1)-sharf+1);
                cnt=0;
                sharf=0;
            }

            ++idx;
            if(compare.rhyme[idx%len]!='#') {
                ++seq;
            }
        }
        
        if(in&&cnt>=m.size()&&compare.rhyme[idx%len]!='#') {
            store.push_back(compare);
        }

        ++rnd;
        ms.clear();
    }

    if(store.size()==0) {
        return "(None)";
    }

    sort(store.begin(),store.end(),[](music a, music b) {
        if(a.et-a.st==b.et-b.st) {
            return a.odr<b.odr;
        }
        return a.et-a.st>b.et-b.st;
    });
    
    answer=store[0].name;

    return answer;
}